hw3/param_model_factory.py

```diff
@@ -74,19 +71,18 @@ def __init__(self, ngram: int):
         self.ngram = ngram
         self._ngram_count = defaultdict(int)
         self._less_one_ngram_count = defaultdict(int)
-        self.avail_tags = set()
+        self.avail_tags = set(START_TAG, STOP_TAG)
 
     @staticmethod
     def _get_ngram(length: int, idx: int, tokens: List[str]) -> Tuple[str]:
-        if idx >= length - 1:
-            return tuple(tokens[i] for i in range(idx - length + 1, idx + 1))
-        return tuple([START_TAG] * (length - idx - 1) + tokens[:idx + 1])
+        return tuple(tokens[i] for i in range(idx - length + 1, idx + 1))
 
     def train(self, train_sentences: List[POSSentence]):
         self._less_one_ngram_count[tuple([START_TAG] * (self.ngram - 1))] = len(train_sentences)
+        self._less_one_ngram_count[tuple([STOP_TAG] * (self.ngram - 1))] = len(train_sentences)        
```

Comment:
this may depend on your implementation of HMM, but I just wonder why do we need multiple END or STOP, I thought only adding one END tag should be ok